pkg/controllers/dashboardapi/settings/settings_test.go

```diff
@@ -0,0 +1,129 @@
+package settings
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	v3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type testCase struct {
+	description     string
+	envVar          *string
+	newSetting      settings.Setting
+	existingSetting *v3.Setting
+}
+
+var settingsETCD = make(map[string]*v3.Setting)
+
+// TestSetAll ensures that SetAll properly updates settings in ETCD. These updates
+// are informed by env vars of the form CATTLE_<SETTING_NAME> and the passed map
+// of settings.Setting objects.
+func TestSetAll(t *testing.T) {
+	settingClient := NewMockSettingClient(gomock.NewController(t))
+	testSettingsProvider := settingsProvider{
+		settings: settingClient,
+	}
+
+	testCases := populateTestCases()
+	settingClient.EXPECT().Get(gomock.Any(), gomock.Any()).DoAndReturn(func(name string, options metav1.GetOptions) (*v3.Setting, error) {
+		val := settingsETCD[name]
+		if val == nil {
+			return nil, errors.NewNotFound(schema.GroupResource{}, name)
+		}
+		return val, nil
+	}).AnyTimes()
+
+	settingClient.EXPECT().Create(gomock.Any()).DoAndReturn(func(setting *v3.Setting) (*v3.Setting, error) {
+		settingsETCD[setting.Name] = setting
+		return setting, nil
+	}).AnyTimes()
+
+	settingClient.EXPECT().Update(gomock.Any()).DoAndReturn(func(s *v3.Setting) (*v3.Setting, error) {
+		settingsETCD[s.Name] = s
+		return s, nil
+	}).AnyTimes()
+
+	settingMap := make(map[string]settings.Setting)
+	for _, testCase := range testCases {
+		settingMap[testCase.newSetting.Name] = testCase.newSetting
+		if testCase.envVar != nil {
+			envKey := settings.GetEnvKey(testCase.newSetting.Name)
+			os.Setenv(envKey, *testCase.envVar)
+			defer os.Unsetenv(envKey)
+		}
+
+		settingsETCD[testCase.newSetting.Name] = testCase.existingSetting
+	}
+	testSettingsProvider.SetAll(settingMap)
+	for _, testCase := range testCases {
+		finalSetting := settingsETCD[testCase.newSetting.Name]
+		failMsg := fmt.Sprintf("test case failed [%s]: %s", testCase.newSetting.Name, testCase.description)
+		assert.Equal(t, finalSetting.Default, testCase.newSetting.Default, failMsg)
+		assert.True(t, finalSetting.Source == "env" == (testCase.envVar != nil), failMsg)
+		if testCase.envVar != nil {
+			assert.Equal(t, *testCase.envVar, finalSetting.Value, failMsg)
+			continue
+		}
+		if testCase.existingSetting != nil {
+			assert.Equal(t, testCase.existingSetting.Value, finalSetting.Value)
+			continue
+		}
+		assert.Equal(t, "", finalSetting.Value)
+	}
+}
+
+func populateTestCases() []testCase {
+	testCases := []testCase{
+		{
```

Comment:
I think these test cases are a good start. There are a few others that also might be good to include:
- Creating a new setting that comes from the env var
- Error on creating a new setting
    - Already exists error?
    - Not already exists error?
- Setting that was not an env var, but now we have an env var for it- https://github.com/rancher/rancher/pull/39453/files#diff-8645d9e3510c3b2f7074d64ef2b316ae3b83a9740f1c60b8a5ab2c92c3483d16L116-L119
- Error when attempting to update the object

These are assuming that we don't want to test the value of fallback. If we end up wanting to do that, then that expands the number of tests a bit more as well.

Reply:
Adding test cases for create returning error and update returning error. Already have a test case for setting with env var that wasn't one before. New setting can't "come from the env var." They should always be created in the file, but added a testcase for a setting that doesn't already exist and configuring with an env var. Added a test case for settings client return AlreadyExists error.